USSF-ORBIT/ussf-portal-cms
# feat: Log in with portal client user

## Description 

This PR replaces the Keystone username/password auth to only use existing Redis sessions created in the portal client. This creates a dependency between running Keystone and running the portal. In order to share sessions, both these env vars need to be defined and use the same value for the two services:
```
SESSION_SECRET
SESSION_DOMAIN
```

The new authentication check logic is:
- Web access is allowed if the user has a session. Only valid sessions are returned.
- Attempting to start a new session from Keystone will throw an error, this should never happen.
- Upon attempt to access Keystone, the user's session cookie is used to pull session data from Redis. If it exists, this data will include the user's SAML data, including their user groups. If no session is found in Redis, the user does not have access (and will have to login to the portal client first).
- If a session is in Redis, the userId attribute is used to look up an existing user in the Keystone DB (Postgres). If no user is found but the user should have access based on their user groups, a new user is created.
- Every time the session is accessed, the user's `isEnabled, isAdmin` flags are synced up with their values from the SAML data and the `syncedAt` timestamp is updated. These flags are both read-only and cannot be edited in the UI.
  - I created a follow-up issue to make this a bit more optimized since right now it updates on every user action involving a session: sc-365

UI permissions are:
- Admin users can view & edit all users (but cannot create or delete)
- Non-admin users can view themselves, and update their display name
- The `userId, isAdmin, isEnabled` fields are read-only and can only be set based on SAML data

Other misc changes:
* Adds e2e docker compose stack for running Cypress tests in CI (which requires the portal client)
* Adds pino logger to Keystone startup
* Also fixes manual deployment script (was using the wrong repository name)

Fixes sc-65

## Review Notes

### Test in AWS dev

This can only be done using your CAC and your actual user. Find out which user groups you belong to, and test that you have the appropriate access by logging into https://dev.ussforbit.us/ and then visiting https://cms.ussforbit.us/

### Test different user roles locally

* Make sure you have the portal client code checked out and up-to-date
* Spin up the services & applications locally using `docker compose -f docker-compose.e2e.yml up -d` (in the CMS repo)
* Log in as the following users using the test SAML IDP and verify their access:
* `user1 / user1pass` (and all other users in users.php) - should not have access
* `cmsuser / cmsuserpass` - should have regular user access, no admin
* `cmsadmin / cmsadminpass` - should have admin access
